python/prepare_release.py

```diff
@@ -0,0 +1,265 @@
+import os
+import uuid
+from typing import Tuple
+
+from github import Github
+
+from .common_tool_methods import *
+from enum import Enum
+
+MULTI_EXTENSION_SQL = "src/test/regress/sql/multi_extension.sql"
+CITUS_CONTROL = "src/backend/distributed/citus.control"
+MULTI_EXTENSION_OUT = "src/test/regress/expected/multi_extension.out"
+CONFIG_PY = "src/test/regress/upgrade/config.py"
+DISTRIBUTED_DIR_PATH = "src/backend/distributed"
+CONFIGURE_IN = "configure.in"
+CONFIGURE = "configure"
+CITUS_CONTROL_SEARCH_PATTERN = r"^default_version*"
+
+MULT_EXT_MAJOR_SEARCH_PATTERN = r"^\s*\d{1,2}\.\d{1,2}devel$"
+MULT_EXT_PATCH_SEARCH_PATTERN = r"^\s*\d{1,2}\.\d{1,2}.\d{1,2}$"
+
+CONFIGURE_IN_SEARCH_PATTERN = "AC_INIT*"
+REPO_OWNER = "citusdata"
+
+
+class ResourceStatus(Enum):
+    INITIAL = 1
+    RELEASE_BRANCH_LOCAL = 2
+    RELEASE_BRANCH_REMOTE = 3
+    UPCOMING_VERSION_LOCAL = 4
+    UPCOMING_VERSION_REMOTE = 5
+    PULL_REQUEST_CREATED = 6
+
+
+def update_release(github_token: str, project_name: str, project_version: is_version(str), main_branch: str,
+                   earliest_pr_date: datetime, exec_path: str, is_test: bool = False,
+                   cherry_pick_enabled: bool = False) -> Tuple[str, str, str, ResourceStatus]:
+    multi_extension_sql_path = f"{exec_path}/{MULTI_EXTENSION_SQL}"
+    citus_control_file_path = f"{exec_path}/{CITUS_CONTROL}"
+    multi_extension_out_path = f"{exec_path}/{MULTI_EXTENSION_OUT}"
+    configure_in_path = f"{exec_path}/{CONFIGURE_IN}"
+    config_py_path = f"{exec_path}/{CONFIG_PY}"
+    distributed_dir_path = f"{exec_path}/{DISTRIBUTED_DIR_PATH}"
+
+    resource_status = ResourceStatus.INITIAL
+
+    project_version_details = get_version_details(project_version)
+    default_upcoming_version = f'{project_version_details["major"]}.{project_version_details["minor"]}.' \
+                               f'{str(int(project_version_details["patch"]) + 1)}'
+    upcoming_version = default_upcoming_version if os.getenv("UPCOMING_VERSION") is None else os.getenv(
+        "UPCOMING_VERSION")
+    upcoming_version_details = get_version_details(upcoming_version)
+    upcoming_patch_version = f'{upcoming_version_details["major"]}.{upcoming_version_details["minor"]}'
```

Comment:
1) upcoming_patch_version -> upcoming_minor_version

2) it seems that rest of the code doesn't need to know about `upcoming_version_details`, can we move those two lines into a function like `get_upcoming_minor_version`

In general, for the places in `update_release` that we are building strings, can we have some nice short functions to improve the readability ?

------

I think any code-blocks starting with a comment can go into separate helper functions. I won't put comments to all such things but I will give some examples for that in below comments.

Reply:
"In general, for the places in update_release that we are building strings, can we have some nice short functions to improve the readability ?"
I think transforming single line operations into methods hurts readibility. Because single line logic will be elsewhere and to understand the operation you shoould go that tline

"I think any code-blocks starting with a comment can go into separate helper functions. I won't put comments to all such things but I will give some examples for that in below comments."
I transformed code into series of methods as you suggest

Reply:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d